winch/codegen/src/isa/x64/masm.rs

```diff
@@ -1163,15 +1174,40 @@ impl MacroAssembler {
 
         Self {
             sp_offset: 0,
+            sp_max: 0,
+            stack_max_use_regions: SmallVec::new(),
             asm: Assembler::new(shared_flags.clone(), isa_flags.clone()),
             flags: isa_flags,
             shared_flags,
             ptr_size: ptr_type.into(),
         }
     }
 
+    /// Add the maximum stack used to a register, recording an obligation to update the
+    /// add-with-immediate instruction emitted to use the real stack max when the masm is being
+    /// finalized.
+    fn add_stack_max(&mut self, reg: Reg) {
+        let open = self.asm.buffer_mut().start_patchable();
+        // NOTE: we write out a value that is not encodable as an imm8, to ensure that we pick the
+        // variation of the x64 add instruction that supports imm32
+        self.asm.add_ir(-129, reg, OperandSize::S32);
```

Comment:
This and the asserts on opcode bytes below feel a little fragile. For example if we change the way immediates are emitted in the future, this might break (we're tying masm's lowering strategy to a specific assumption about its output here). Also the REX byte is pinned by an assert, and might change if the regalloc picks a register in the other half of the 16-reg bank (!). Perhaps what we could do instead is something like:

- Have a `PatchableValueLoad` abstraction (either an `MInst` case or a little struct here that does its own encoding, to nail down the length-invariance); ensure that it always produces machine code of the same length. `::new(value: u32, reg: Reg)`, `::update_value(value: u32)` and `::emit_to_slice(&mut [u8])` or something like that.
- Use it here to initially put a placeholder in; we can avoid using a magic value and magic knowledge about immediate encoding then. Perhaps we can even write a wrapper `::emit_to_buffer(&mut MachBuffer) -> PatchRegion`.
- Later, patch using this; perhaps with `::emit_to_region(&mut MachBuffer, PatchRegion)`.

Thoughts?

Reply:
I agree that it's fragile as-is, and something a bit more robust would be good here.

Adding a special `PatchableValueLoad` type sounds good to me, as long as you don't mind the more specific "add patched immediate to register" behavior. I wanted to avoid using an additional register here, so that the sequence of loading the stack lower bound and adding the immediate for the final stack value could be done with only the scratch register.

I'll experiment with adding a `PatchableAddRegImm` struct in the x64 backend in cranelift, as that would make it easy to reuse logic from the existing emit implemetation for add.

Reply:
Would you mind taking a look at the changes in x64/asm.rs? I had to copy some details over from cranelift's emit module for rex encoding, but I believe that this approach should be much more resilient: it doesn't assume anything about how cranelift would encode an instruction; it records the offset into the patch region where the immediate starts, so it should also be resilient to local changes.

The downside is that it brought a lot of details into the asm module, but I think that's somewhat unavoidable at this  point.